### ISSUE TITLE
set_key_transform makes set_type ignored

### ISSUE THREAD

**Issue by addicted2sounds** [Netflix/fast_jsonapi#293]
as I can see from the code `set_key_transform` makes `set_type` to be ignored:

```
      def reflected_record_type
        return @reflected_record_type if defined?(@reflected_record_type)

        @reflected_record_type ||= begin
          if self.name.end_with?('Serializer')
            self.name.split('::').last.chomp('Serializer').underscore.to_sym
          end
        end
      end

      def set_key_transform(transform_name)
        mapping = {
          camel: :camelize,
          camel_lower: [:camelize, :lower],
          dash: :dasherize,
          underscore: :underscore
        }
        self.transform_method = mapping[transform_name.to_sym]

        # ensure that the record type is correctly transformed
        set_type(reflected_record_type) if reflected_record_type
      end
```

Do I understand correctly that I can't set my own record type if I want to use `key_transform`?

**Comment by manojmj92** [comment]
@addicted2sounds I don't think this is right. 
Because : https://github.com/Netflix/fast_jsonapi/blob/master/lib/fast_jsonapi/object_serializer.rb#L161

clearly specifies to run `run_key_transform ` on the specified `record_type` you have specified using `set_type`

**Comment by addicted2sounds** [comment]
@manojmj92 As I understand then, I should place set_type only after set_key_transform. `set_type` is then invoked twice:
1. in set_key_transform: https://github.com/Netflix/fast_jsonapi/blob/5ff3fa97dae4b6be51bf8ec315cb427667c4ce96/lib/fast_jsonapi/object_serializer.rb#L145
2. when I actually call `set_type`

**Comment by manojmj92** [comment]
Yes, that is right. Calling `set_type` after `set_key_transform` performs transformation on the input twice, but the result will be same in both the first pass and second pass.

However, calling `set_key_transform` after `set_type` also performs transformation on the input twice, but the result of the first pass will be different from the second pass.

But whichever order you call them makes no difference to the end result (ie, after both passes the end result remains same in both cases)

While this duplication of calling twice can be avoided, we have it in place because we do not specify any hard and fast rule about the order in which these methods should be called.

Remaining true to the beauty of Ruby, I think we should continue giving this flexibility of choosing their own order to users, and even if it leads to some minor duplication like this, it is OK, in my opinion :)
